wordpress-mobile/WordPress-iOS
# Display status of posts with autosave revision available (part 1)

Posts on the post list screen now show a status label for posts that have an autosave revision available. A later PR will add the dialog and support for editing the autosaved revision.

⚠️`Podfile` and `Podfile.lock` are temporarily updated to point to the corresponding WordPressKit feature branch.

Issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/12141

### Visual Changes

The second post in the screenshots below has an autosave revision available.

| Regular Layout | Compact Layout |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/1898325/67648943-f154b280-f915-11e9-88f0-046046092c85.png" width="250"> | <img src="https://user-images.githubusercontent.com/1898325/67649016-26f99b80-f916-11e9-88d5-15beeaf58426.png" width="250"> | 

### Outstanding Questions
- Is autosave supported via XML RPC? This PR adds support for the REST API only
- Autosave data includes the author ID - does this mean a user can fetch another user's autosave?
- Is the autosave preview URL used? I noticed it added on Android

### Observations
- The status text is sometimes truncated due to its length in certain situations (e.g. on compact layout)
- Autosave applies to Published as well as Scheduled posts
- The status label uses the color, `.warning(.shade40)` which should be the same as Android
- Both `BasePost` as well as its subclass `AbstractPost` are stored in Core Data - the autosave properties were added to `BasePost` because it contained the existing title, excerpt, and content
- The Core Data migration here is automatic because only new properties are added
- Both `- updatePost:withRemotePost:` and `- remotePostWithPost` are updated to handle the new autosave properties

### Testing
1. On WordPress.com, create a new post or open an existing post for editing
2. Set the title to “Hello, world!” and the body to “One.”
3. Publish the post
4. Re-open the post and add a new line, “Two.” – then close the browser tab without saving
5. Again, re-open the post and notice the following message -  “The backup of this post in your browser is different from the version below. Restore the backup”  – this signifies that an autosave was made
6. On the WordPress for iOS app, under the My Sites tab, navigate to the list of posts containing the newly created post
7. Observe that there is now a status label ("You've made unsaved changes to this post") indicating that an autosave version is available
